### ISSUE TITLE
Umple template false positive cycle detection

### ISSUE THREAD

**Issue by Diusrex** [umple/umple#736]
In general, the cycle detection will correctly determine if a template causes a cycle (although there may be some cases like issue #735 where this is not true).

However, for the template

```
class A
{
    baseTemplate <<! Base Template !>>

    template <<! <<@ baseTemplate >> <<@ baseTemplate >> !>>
    template2 <<! <<@ template >> !>>

    templ

    emit noCycle()(template); // Fine, no cycle
    emit fails()(template2); // Incorrectly says there is a cycle
};
```

it incorrectly determines that template2 contains a cycle.

This appears to only occur when one or more children of a template include the same template multiple times (where a child would be a template that is eventually included). It does not appear to occur when multiple children include the same template, so long as each template only includes a template once.

Note that, for a template `A`, and includes a template `B`, where `A` eventually includes `B`, there is no false positive:

```
class A
{
    templateA  <<! <<@otherTemplateWithB >> <<@templateB>> !>>
    otherTemplateWithB <<! <<@templateB >> !>>
    templateB <<! Base Template !>>

   emit noCycle()(templateA); // Again, is fine
}
```

**Comment by TimLethbridge** [comment]
@ahmedvc Just merged pull request #738  that may solve this issued also #735 and #732 . Would @Diusrex comment and if the problem is solved, then close this issue please.

**Comment by Diusrex** [comment]
I can confirm that this issue has been fixed with the changes that they submitted.
